KEP-4639: adding OCI VolumeSource (#4642)
kubernetes/enhancements: keps/sig-node/4639-oci-volume-source/README.md

well as the [existing list] of feature gates.

[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
-->

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: OCIVolume
  - Components depending on the feature gate:
    - kube-apiserver (API validation)
    - kubelet (volume mount)

###### Does enabling the feature change any default behavior?

<!--
Any change of default behavior may be surprising to users or break existing
automations, so be extremely careful here.
-->

Yes, it makes the new `VolumeSource` API functional.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

<!--
Describe the consequences on existing workloads (e.g., if this is a runtime
feature, can it break the existing applications?).

Feature gates are typically disabled by setting the flag to `false` and
restarting the component. No other changes should be necessary to disable the
feature.

NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
-->

> Yes, by disabling the feature gate. Existing workloads will not be affected by

Reviewer: Can we get some more elaboration?  Which processes need to be restarted in which order to clear the old volume sources?  I'm betting all the kubelets have to be restarted and all pods evicted?  Before or after the kube-apiserver?

Reviewer 2: I added more details to that point. In theory it would be enough to disable the feature on a kubelet + restart the workload to make the volume disappear while keeping the API validation intact. 

If we disable the feature at the API level, then the fields will not be set and the kubelet cannot do the implementation part.

the change.

###### What happens if we reenable the feature if it was previously rolled back?

> It will make the API functional again.

Reviewer: When re-enabled, will the latent volume sources become active again?  Will we end up in a mismatch for pods under a deployment or daemonset, where some will have a new volume source and some will not (I bet we will).

Reviewer 2: Yes, that's true. I'm wondering if we should reject pods which have the field set (API enabled) but the kubelet feature gate is not available. This would mean that in a diverging scenario of a deamonset only a subset of pods can be started while others cant because the kubelet feature is disabled).

-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

The proposed enhancement adds a new `VolumeSource` to Kubernetes that supports OCI images and/or OCI artifacts.
This allows users to package files and share them among containers in a pod without including them in the main image,
thereby reducing vulnerabilities and simplifying image creation.

While OCI images are well-supported by Kubernetes and CRI, extending support to
OCI artifacts involves recognizing additional media types within container
runtimes, implementing custom lifecycle management, resolution of artifact
registry referrers use pattern for artifacts, and ensuring appropriate
validation and security measures.

## Motivation

Supporting OCI images and artifacts directly as a `VolumeSource` allows
Kubernetes to focus on OCI standards as well as allows to store and distribute
any content using OCI registries. This allows the project to grow into use cases
which go beyond running particular images.

### Goals

> - Introduce a new `VolumeSource` type that allows mounting OCI images and/or artifacts.

Reviewer 3: Aren't we saying though that we're going to (in this kep)

1. Mount OCI images
2. Delegate any unknown OCI media types to the runtime (thus shifting the burden to the container runtime?)

There seems to be some risks to 2 - how would we define conformance?  I would definitely expect an intree volume mount to clearly define conformance behavior for 1, but I can't quite describe it for 2.  However, the counter argument is that we don't define any of this for container images today (we simply conformance test on known OCI image types), and the exact same behavior would happen - as long as a filesystem exists after the OCI pull action is resolved, kube doesn't care.

So that to me feels like the bar - "a VolumeSource type that allows mounting OCI images and any future OCI artifacts that can be expanded to a file or directory representation"

Reviewer 4: It may be easiest to just put other types of oci artifacts out of scope for now. When other oci artifact types that may make sense to mount as a filesystem show up, we consider them in a future kep?

Reviewer 5: We could specify how the directory structure will look like. There are some considerations for how volumes will be shared with SELinux labels and we may need to overlay layer them but as long as final directory structure is the same across runtimes, that should work.

Reviewer 2: I added a point to:

> Providing a runtime guideline of how artifact files and directories should be mounted.

- Simplify the process of sharing files among containers in a pod.

> ### Non-Goals

Reviewer 3: We might want to include a non-goal of mounting thousands of images and artifacts, unless we have a case for that (the kube pod spec does not scale to mounting more than 100s of things).

Reviewer 5: Agree :+1:

Reviewer 2: Added

- This proposal does not aim to replace existing `VolumeSource` types.

> - This proposal does not address other use cases for OCI objects beyond file sharing among containers in a pod.

Reviewer 3: I think we should be precise and say "file or directory sharing among containers in a pod".  Directory is the most common, but standard volumes are expected to support both file and directory semantics today.

Reviewer 2: Fixed as suggested :+1:

## Proposal

We propose to add a new `VolumeSource` that supports OCI images and/or artifacts. This `VolumeSource` will allow users to mount an OCI object
directly into a pod, making the files within the image accessible to the containers without the need to include them in the main image and to be able to host them in OCI compatible registries.

> ### User Stories (Optional)

Reviewer 3: Sally mentioned this in slack as potentially being out of scope, but i don't see any story that would require efficient inline writes mentioned (overlay).  Do we have any?  

One I can think of is a build cache directory that is populated from an image (efficient distribution of precompiled content) where subsequent writes happen to the cache dir in an overlay.

For this to be useful, the amount of content has to be large enough that copy from the volume mount to a scratch filesystem would be slow (unless we can make reflinking work consistently???), and the content in the image is sufficiently stable that the overhead of creating the image is low, yet have a high hit rate.  Some build / cache systems might achieve that, but not many.

Otherwise, we should put that as a non-goal of this KEP up above.

Reviewer 2: @sallyom do we want to add another story about that?

Author: I see the use case, but still I'd like to keep this simple - the option of a writeable overlay can be added later if the community finds they need it - the driving use-case for this is serving models - my thinking (wrt AI) is that mounting models as OCI images is for serving/inferencing (read only) while with training & fine-tuning models you wouldn't mount the files in this way. I would prefer to add this as a non-goal of this KEP.

#### Story 1

As a Kubernetes user, I want to share a configuration file among multiple containers in a pod without including the file in my main image, so that I can
minimize security risks and image size. 

Beside that, I want:
- to package this file in an OCI object to take advantage of OCI distribution.
- the image to be downloaded with the same credentials that kubelet using for other images.

> - to be able to use image pull secrets when downloading the image if an image is from the registry that requires image pull secrets.

Reviewer 3: What none of these stories specifically clarify or not is what their lifecycle needs are - i.e. do we need to support changing the volume image source and have it update?  What happens with pull policy?

I would expect either a story that talks about whether updates of images is needed, and if not I think we should call it out in non-goals.

Reviewer 2: I added this to the first story by using a moving tag like `latest`.

#### Story 2

As a DevOps engineer, I want to package and distribute binary artifacts using
OCI image and distribution specification standards and mount them directly into
my Kubernetes pods, so that I can streamline my CI/CD pipeline. This allows me to
maintain a small set of base images by attaching the CI/CD artifacts to them.
Beside that, I want to package the artifacts in an OCI object to take advantage
of OCI distribution.

#### Story 3

> As a data scientist, MLOps engineer, or AI developer, I want to mount large language models or machine learning models in a pod alongside a model-server, so that I can efficiently serve the models

Reviewer 2: Switched it to "model weights" as suggested.

   - A container image that conforms to the Open Container Initiative (OCI) Image Specification.
     It includes a filesystem bundle and metadata required to run a container.
   - Consists of multiple layers (each layer being a tarball), a manifest (which lists the layers), and a config file
     (which provides configuration data such as environment variables, entry points, etc.).
   - **Use Case:** Used primarily for packaging and distributing containerized applications.

**OCI Artifact ([guidance](https://github.com/opencontainers/image-spec/blob/main/artifacts-guidance.md)):**
   - An artifact describes any content that is stored and distributed using the OCI image format.
     It includes not just container images but also other types of content like Helm charts, WASM modules, machine learning models, etc.
   - Artifacts use the same image manifest and layer structure but may contain different types of data
     within those layers. The artifact manifest can have media types that differ from those in standard container images.
   - **Use Case:** Allows the distribution of non-container content using the same infrastructure and tools developed for OCI images.

**OCI Object:**

>    - Umbrella term encompassing both OCI images and OCI artifacts. It represents any object that conforms to the OCI specifications for storage and distribution.

Reviewer 3: Similar to comment up above, the assumption is that "all objects can be converted to a file or filesystem representation by an OCI container runtime"?  If so, that's a good framing that might belong in notes / goals (all OCI objects that can be converted to files or directories can be mounted).

### Risks and Mitigations

> - **Security Risks:** Allowing direct mounting of OCI images introduces potential attack vectors. Mitigation includes thorough security reviews and

Reviewer 3: Path traversal attacks against mounts are the #1 concern for me in any volume implementation.  I would expect us to articulate either what we defend against or whether it's not worthy of defense here.

Reviewer 2: Added apoint about that. Re-using the existing container storage implementations for most operations should be a good mitigation.

  limiting access to trusted registries. Limiting to OCI artifacts (non-runnable content) or read-only mode may lessen the security risk.
- **Compatibility Risks:** Existing webhooks watching for the images used by the pod using some policies will need to be updated to expect the image to be specified as a `VolumeSource`.
- **Performance Risks:** Large images or artifacts could impact performance. Mitigation includes optimizations in the implementation and providing
  guidance on best practices for users.

## Design Details

The new `VolumeSource` will be defined in the Kubernetes API, and the implementation will involve updating components (CRI, Kubelet)
to support this source type. Key design aspects include:

- API changes to introduce the new `VolumeSource` type.
- Modifications to the Kubelet to handle mounting OCI images and artifacts.
- Handling image pull secrets and registry authentication.
- Reuse existing logic from ConfigMaps to determining the file location on the host filesystem.

The following code snippet illustrates the proposed API change:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: example-pod
spec:

>   volumes:

Reviewer 3: I'm coming in later and trying to catch up, so apologies if this is well covered.  But it's not clear from the kep how we are separating out the source (and what form it is in: filesystem, multiple files, or single file) and how a user disambiguates it.

* Conceptually we are already mounting images as filesystems for containers.  
* Volume sources can be both filesystems (simplest form is a directory) or a single file (e.g. secret mount)
* Artifacts may be filesystems, multiple-files with names, single file with a name, or potentially a single file without a name?

Therefore we are talking about the following actions

1. Mount image as filesystem (no qualification beyond image)
2. Mount image as file (somehow point to a file in the image via a path spec)
3. Mount image as directory (somehow point to a directory in the image via a path spec)
4. Mount artifact as filesystem (the artifact must have a "name" that can be represented in a directory)
5. Mount artifact as file (artifact is a singular file, otherwise the user has to disambiguate which artifact?)
6. Mount artifact as directory (artifact either has multiple entries that have names, or is a compressed filesystem, or has a single entry that has a name)

?  What isn't clear to me is how the user specifies 2 and 3, and then whether artifacts are guaranteed to have names (which determines what has to be provided by the user for 5 and 6).

Reviewer 6: Imagine an OCI artefact that is an unnamed raster image. Eventually, the thing we design here might want to try to mount that.

However, I think we have an option to intentionally make that example case a non-goal for this KEP (and instead suggest using an alternative, such as a CSI driver). Any CSI driver needs to know how to turn the raster data into a volume, even if it's as simple as “take the raw binary data and make it available as a file called _x_”.

There are two obvious places to put a boundary (at least, that's how I see it):
- we'd try to handle all things that can be fetched from an OCI registry, and the boundary is whether we want to just mount it into a Pod or run it as a container
- we'd (eventually?) like to handle lots of things that can be fetched from an OCI registry, and the boundary is whether we want the CRI-level container runtime to do the fetching or not. Container runtimes only know about fetching container-like images / we can't assume they have additional capabilities.

---

Overall, I'd like to do the simplest thing we can live with and not find ourselves painted into an API corner. Taking the single-file case out of the KEP looks good, especially if it gets us to a point where we can merge it as provisional.

Reviewer 7: > Artifacts may be filesystems, multiple-files with names, single file with a name, or potentially a single file without a name?

At the lowest level artifacts are basically "ship things that aren't tarballs wrapped with the same JSON as OCI container images".

https://github.com/opencontainers/image-spec/blob/main/manifest.md#guidelines-for-artifact-usage

The WASM OCI artifact type docs are a good example: https://tag-runtime.cncf.io/wgs/wasm/deliverables/wasm-oci-artifact/

The use case of "mount AI model weights stored as an OCI artifact" to me seems most useful...it's generic *architecture independent* content so packing it as a container introduces logistical overhead (need to "build" it for multi-arch scenarios even though it isn't).

I actually don't know if there's an existing spec for it...would be useful if there was!

> Container runtimes only know about fetching container-like images / we can't assume they have additional capabilities.

At least IMO if something like this ends up in core kubelet it'd also get exposed by `podman play kube` etc.

Reviewer 2: I added a section "Filesystem Representation" covering the above point. Please continue the discussion there if there is anything left open or I missed.

  - name: oci-volume
    oci:
      image: "example.com/my-image:latest"

>       pullPolicy: IfNotPresent

Reviewer 3: We need to specify when the pull policy is triggered, because this is a pod level construct now and the previous pull policy was container level.  

I.e. containers repull when the container is recreated, but pullPolicy: Always on a volume spec that is mounted into two different containers is not the same as containers and needs defined semantics.

Reviewer 2: I added more details about that point in 3f72dced

  containers:
  - name: my-container
    image: busybox
    volumeMounts:
    - mountPath: /data
      name: oci-volume
```

### Kubelet and Container Runtime Interface (CRI) support for OCI artifacts

Kubelet and the Container Runtime Interface (CRI) currently handle OCI images. To support OCI artifacts,
potential enhancements may be required:

**Extended Media Type Handling in the container runtime:**
   - Update container runtimes to recognize and handle new media types associated with OCI artifacts.
   - Ensure that pulling and storing these artifacts is as efficient and secure as with OCI images.

**Lifecycling and Garbage Collection:**

>    - Reuse the existing kubelet logic for managing the lifecycle of OCI objects.

Reviewer 8: we should be able to do a pull with unpack on runpodsandbox add a lease at the pod level for the mountable snapshot(s) layer(s) for these pod images/artifacts, that will be attached later to the containers rootfs per source/target path requests...

From there removing the image should be ok.. as these pod's oci volume mounts for each container would already be prepared.

>    - Reuse the existing kubelet logic for managing the lifecycle of OCI objects.

Reviewer 2: Yes, this is an additional check in the garbage collection. Added a note about that.

- The retrieval of available secrets for a pod:
  https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/kubelet_pods.go#L988

Specifying a `pullPolicy` be supported in the same way as for the existing
`imagePullPolicy` API.

For registry authentication purposes the same logic will be used as for the
container image.

#### CRI

The CRI API is already capable of managing container images [via the `ImageService`](https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L146-L161).
Those RPCs will be re-used for managing OCI artifacts, while the [`ImageSpec`](https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L798-L813)
as well as [`PullImageResponse`](https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L1530-L1534)
will be extended to mount the OCI object to a local path:

```protobuf

// ImageSpec is an internal representation of an image.
message ImageSpec {
    // …

    // Indicate that the OCI object should be mounted.
    bool mount = 20;

    // SELinux label to be used.
    string mount_label = 21;

> }

Reviewer 9: Are there any `fsGroup` or `uid` based permissions to consider?

Reviewer 2: I don't think so, that's handled by the runtime providing the mount. What would be the use case?

Reviewer 9: Is there a upper limit to image sizes? My concern is - if there are large images with lots of files, runtime is recursively both changing selinux perms and file based permissions and then we will hear complaints from customers that pods that are using these images are failing to start. 

We have worked around this via APIs in kubernetes to some extent. (See selinux and fsgroup related KEPs). Are these solved problems for CRI? Will these volume types use any of existing APIs that control fsgroup and selinux handling in pods from storage POV? cc @jsafrane

Reviewer 7: What use cases would there be for a custom label?  I think the mount should always use the same label type as that for the base pod image.

> Is there a upper limit to image sizes? My concern is - if there are large images with lots of files, runtime is recursively both changing selinux perms and file based permissions and then we will hear complaints from customers that pods that are using these images are failing to start.

Note that these are not *mutable* volumes, so there's no need to chown anything ever. I think in practice the default would be that the files in the volume are root owned and world readable, that's it.

Reviewer 2: @cgwalters yes, my idea was to use the `mount_label` field in the CRI inherited from the pod. Then use `mount -o context=…` in the runtime (there is an API) to read-only mount with the target label.

Reviewer 9: Will this `VolumeSource` support `subPaths` ? If yes - using `-o context` may not always work....

Reviewer 9: It exists as an API in pod's `volumeMount` field. I am not sure if there is a way to block it. What will happen if different containers in the pod are trying to mount different `subPath`'s from the image?

Reviewer 2: > It exists as an API in pod's volumeMount field. I am not sure if there is a way to block it.

The API validation should be able to do that: https://github.com/kubernetes/kubernetes/blob/efef32652af0af08a0b9c9bc547a4dce4a95f9f5/pkg/apis/core/validation/validation.go#L3613

> What will happen if different containers in the pod are trying to mount different subPath's from the image?

That will be forbidden.

feature cannot be used. The kubelet will verify if the returned `mountpoint`
actually exists on disk to check the feature availability, because Protobuf will
strip the field in a backwards compatible way for older runtimes. Pods using the
new `VolumeSource` combined with a not supported container runtime version will
fail to run on the node.

For security reasons, volume mounts should set the [`noexec`] and `ro`
(read-only) options by default.

##### SELinux

Traditionally, the container runtime is responsible of applying SELinux labels
to volume mounts, which are inherited from the `securityContext` of the pod or
container. Relabeling volume mounts can be time-consuming, especially when there
are many files on the volume.

If the following criteria are met, then the kubelet will use the `mount_label`
field in the CRI to apply the right SELinux label to the mount.

- The operating system must support SELinux
- The Pod must have at least `seLinuxOptions.level` assigned in the
  `PodSecurityContext` or all volume using containers must have it set in their
  `SecurityContexts`. Kubernetes will read the default user, role and type from

>   the operating system defaults (typically `system_u`, `system_r` and `container_t`).

Reviewer 9: This seems like yet another way of accomplishing - https://github.com/kubernetes/enhancements/issues/1710 but it doesn't fit with existing design.

Reviewer 2: Moving that discussion into https://github.com/kubernetes/enhancements/pull/4642#discussion_r1636806425

This is a good solution for restoring a volume from a snapshot or initializing a volume with data from a database backup. However, it does not
address the desire to use OCI distribution, versioning, and signing for mounted data.

The proposed in-tree OCI VolumeSource provides a direct and integrated approach to mount OCI artifacts, leveraging the existing OCI
infrastructure for packaging, distribution, and security.

### Custom CSI Plugin

See [https://github.com/warm-metal/container-image-csi-driver](https://github.com/warm-metal/container-image-csi-driver)

An out-of-tree CSI plugin can provide flexibility and modularity, but there are trade-offs to consider:

 - Complexity of managing an external CSI plugin. This includes handling the installation, configuration, and updates of the CSI driver, which adds
   an additional operational burden. For a generic, vendor-agnostic, and widely-adopted solution this would not make sense.
 - Supporting the image pull secrets as well as credentials provider will be tricky and needs to be reimplemented with the separate API calls.

>  - External CSI plugins implement their own lifecycle management and garbage collection mechanisms,

Reviewer 10: Other considerations:

- kubelet has max parallel image pull constant to maintain the reasonable load on a disk and network. This will not be respected by CSI driver and the only point of integration may be if we move this constant down to runtime
- kubelet has GC logic that is not cleaning up images immidiately in case they will be reused. Also GC logic has it's own thresholds and behavior on eviction. It will be nice to have those integrated
- kubelet exposes metrics on image pulls and we have KEP in place to improve it even further. Having CSI exposing those metrics will require customer to integrate with one more source of data

It is all implementeable in CSI plugin, just adds coplexity for the benefit of being out of tree.

I think the biggest advantage of out of tree will be if it will be super rare use case. But I think a lot of people will choose the image registries as a versioned file distribution mechanism if this feature will be available.
-

Reviewer 2: Thank you, I added the points.

      the container storage prevent such issues.
- **Compatibility Risks:** Existing webhooks watching for the images used by the pod using some policies will need to be updated to expect the image to be specified as a `VolumeSource`.
- **Performance Risks:** Large images or artifacts could impact performance. Mitigation includes optimizations in the implementation and providing
  guidance on best practices for users.

## Design Details

The new `VolumeSource` will be defined in the Kubernetes API, and the implementation will involve updating components (CRI, Kubelet)
to support this source type. Key design aspects include:

- API changes to introduce the new `VolumeSource` type.
- Modifications to the Kubelet to handle mounting OCI images and artifacts.
- Handling image pull secrets and registry authentication.
- Reuse existing logic from ConfigMaps to determining the file location on the host filesystem.

The following code snippet illustrates the proposed API change:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: example-pod
spec:
  volumes:
  - name: oci-volume
    oci:
      image: "example.com/my-image:latest"

>       pullPolicy: IfNotPresent

Reviewer 11: Would the [AlwaysPullImages](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#alwayspullimages) admission plugin also set this field to `Always` like it does for container images?   For multi-tenant clusters I think you would want the AlwaysPullImages admission plugin (or another one) to handle this.

Reviewer 2: Good point, I'll add this to the "Pull Policy" section.

     registry/repository. The new referrers API allows for retrieving these
     artifacts.

**Validation:**
   - Extend validation and security checks to cover new artifact types.

**Storage Optimization in the container runtime:**
   - Develop optimized storage solutions tailored for different artifact types,
     potentially integrating with existing storage solutions or introducing new mechanisms.

#### kubelet

While the container runtime will be responsible of pulling and storing the OCI
objects in the same way as for images, the kubelet still has to manage the full
life-cycle of them. This means that some parts of the existing kubelet code can
be reused, for example:

- The logic how to ensure that an image exists on the node:
  https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/images/image_manager.go#L102
- The retrieval of available secrets for a pod:
  https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/kubelet_pods.go#L988

##### Pull Policy

While the `imagePullPolicy` is working on container level, the introduced
`pullPolicy` is a pod level construct. This means that we can support the same

> values `IfNotPresent`, `Always` and `Never`, but will only pull once per pod.

Reviewer 11: Is the actual OCI manifest or config digest going to be recorded somewhere in the pod status or the volume itself so it is visible (via `kubectl`)?  We have this property with regular images and it is nice when pulling by tags to see what was actually pulled when the pod started.

Reviewer 2: I added a point about that in the "CRI" section. Runtimes have to track the link between sandboxes and mounted OCI volumes so that they can provide that information to the kubelet. This is required for the kubelet image garbage collection but can be (later on) used by clients like kubectl

before the container gets started.

If users want to re-pull artifacts when referencing moving tags like `latest`,
then they need to restart / evict the pod.

##### Registry authentication

For registry authentication purposes the same logic will be used as for the
container image.

#### CRI

The CRI API is already capable of managing container images [via the `ImageService`](https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L146-L161).
Those RPCs will be re-used for managing OCI artifacts, while the [`ImageSpec`](https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L798-L813)
as well as [`PullImageResponse`](https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L1530-L1534)
will be extended to mount the OCI object to a local path:

```protobuf

// ImageSpec is an internal representation of an image.
message ImageSpec {
    // …

    // Indicate that the OCI object should be mounted.
    bool mount = 20;

    // SELinux label to be used.
    string mount_label = 21;
}

message PullImageResponse {
    // …

    // Absolute local path where the OCI object got mounted.

>     string mountpoint = 2;

Reviewer 8: mounted on a pull image request `with unpack` .. for a particular `pod image:count:tag (pid) + runtime handler`.. hmm

Reviewer 8: if there are two pods with the same pull.. there could be two snapshots and could be two different versions if latest, or if the same sha could be optimized.. if two different runtime handlers there could be two snapshots.. 

Somehow we need to lock in the mount point and I'm not sure that is during pull image as that is not a part of the pod lifecycle.

This is a part of why on a container we may end up pulling at least a portion of the image again on create container even after the image has been ensured to be present by Kubelet. This was mentioned by Clayton in this kep discussion.

Reviewer 8: I wish we could split runpodsandbox into create and start.. Until then my recommendation is pass in the pod image/artifact mount(s) in the runtpodsandbox response

Reviewer 8: if there are two pods with the same pull.. there could be two snapshots and could be two different versions if latest, or if the same sha could be optimized.. if two different runtime handlers there could be two snapshots..

Somehow we need to lock in the mount point and I'm not sure that is during pull image as that is not a part of the pod lifecycle.

This is a part of why on a create container we check resolve, and if it can't resolve kubelet will have to ensure again, even after the image has been ensured a first time to be present by Kubelet. This was mentioned by Clayton in this kep discussion. At one point in history, we would pull the image a second time in create container vs ensure we can resolve and fail the create if resolve fails. https://github.com/containerd/containerd/blob/main/internal/cri/server/container_create.go#L107

I wish we could split runpodsandbox into create and start.. Until then my recommendation is return the pod image/artifact mount(s) in the runpodsandbox response

Reviewer 2: Let's discuss that topic in real time (pinged you on slack including @haircommander and @mrunalp)

Reviewer 2: The current design proposal still pulls on the `PullImage` RPC, but then passes the `Mount` as part of the `CreateContainer` call. This way we can fail if a mount does not exist and can ensure to have the same referring bind mounts on container creation.

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: example-pod
spec:
  volumes:
  - name: oci-volume
    oci:
      image: "example.com/my-image:latest"
      pullPolicy: IfNotPresent
  containers:
  - name: my-container
    image: busybox
    volumeMounts:
    - mountPath: /data
      name: oci-volume
```

### Kubelet and Container Runtime Interface (CRI) support for OCI artifacts

Kubelet and the Container Runtime Interface (CRI) currently handle OCI images. To support OCI artifacts,
potential enhancements may be required:

**Extended Media Type Handling in the container runtime:**
   - Update container runtimes to recognize and handle new media types associated with OCI artifacts.
   - Ensure that pulling and storing these artifacts is as efficient and secure as with OCI images.

**Lifecycling and Garbage Collection:**
   - Reuse the existing kubelet logic for managing the lifecycle of OCI objects.
   - Extending the existing image garbage collection to not remove an OCI volume
     image if a pod is still referencing it. 

**Artifact-Specific Configuration:**
   - Introduce new configuration options to handle the unique requirements of different types of OCI artifacts.

> **Artifacts as Subject Referrers:**

Reviewer 12: It's unclear to me how this new volume type will interact with the referrers API. Is the idea to also mount artifacts that are referenced by the `image` field in the volume source? Can you add an example?

Reviewer 2: This is mostly for future graduations and should be handled more in runtimes rather than Kubernetes. For example SBOMs can be referenced to verify OCI artifacts or images. But that's something we can discuss later on, I clarified that in the proposal.

     the same mechanisms will be reused. This ensures consistency and reduces complexity, as both adhere to the same OCI image format.
   - **Existing Lifecycle Management and Garbage Collection:** Kubernetes has efficient lifecycle management and garbage collection mechanisms for
	 volumes and container images. The in-tree OCI VolumeSource will utilize these existing mechanisms.

2. **Integration with Kubernetes:**
   - **Optimal Performance:** Deep integration with the scheduler and kubelet ensures optimal performance and
     resource management. This integration allows the OCI VolumeSource to benefit from all existing optimizations and features.
   - **Unified Interface:** Users interact with a consistent and unified interface for managing volumes, reducing the learning curve and
     potential for configuration errors.

3. **Simplified Maintenance and Updates:**
   - **Core Project Maintenance:** In-tree features are maintained and updated as part of the core project. It makes sense
     for widely-used and vendor agnostic features to utilize the core testing infrastructure, release cycles, and security updates.

> 

Reviewer 13: The one obvious alternative that was not even listed here is using shared volumes inside a single pod. Was this considered? What were the reasons for a new volume source?

Author: Yup, this is what prompted the proposal - I was annoyed because I wanted to mount  an LLM as a separate image rather than combine an LLM + model-server into 1 image (for modularity/easily plugging in different models). In order to do this, I had to package the LLM in an image with a shell in its base layer. Then, an init container to copy the LLM (using the linux `install` command) from the model image to a volume shared with the model-server container. In order to remove the need for a shell & init container, I propose a new OCI VolumeSource. That alternative _should_ be explicitly stated here, I'll add it!

Reviewer 14: If you mount and ORAS artifact you can do away with your LLM data also needing a filesystem.

Reviewer 2: I assume this conversation is resolved with https://github.com/kubernetes/enhancements/pull/4642/commits/e5013dc14aba730b4efad49453fb821d0ecbd551, feel free to reopen if I've missed anything.

Reviewer 13: This still holds, see my other comment.

potential enhancements may be required:

**Extended Media Type Handling in the container runtime:**
   - Update container runtimes to recognize and handle new media types associated with OCI artifacts.
   - Ensure that pulling and storing these artifacts is as efficient and secure as with OCI images.

**Lifecycling and Garbage Collection:**
   - Reuse the existing kubelet logic for managing the lifecycle of OCI objects.
   - Extending the existing image garbage collection to not remove an OCI volume
     image if a pod is still referencing it. 

**Artifact-Specific Configuration:**
   - Introduce new configuration options to handle the unique requirements of different types of OCI artifacts.

**Artifacts as Subject Referrers:**
   - Introduce new refer to image and filter for artifact type criterion/options for to be mounted artifact(s).
   - Certain types of OCI artifacts include a subject reference. That reference
     identifies the artifact/image for which this artifact refers. For example a
     signature artifact could refer to a platform index, for certifying the
     platform images, or to an SBOM artifact that refers to a platform matched

>      image. These artifacts may or may not be located on the same
>      registry/repository. The new referrers API allows for retrieving these
>      artifacts.

Reviewer 15: While it is possible the subject could reference content in a different repository, registry behavior may vary (a GC could delete the data as an untagged manifest), and the referrers API will only respond with content found in the requested repository.

```suggestion
     image. These artifacts may or may not be located on the same
     registry/repository. The new referrers API allows for discovering
     artifacts from a requested repository.
```

Reviewer 8: ```suggestion
     image. These artifacts may or may not be located on the same
     registry/repository. The new referrers API allows for discovering subject referring artifacts in an OCI v1.1 compatible registry. These subject referring artifacts may then be downloaded for further processing at the client.
```

Reviewer 15: @mikebrow I believe your edit leaves the false impression that the referrers API would respond with artifacts pushed to a different repository or registry. Is that intentional? If so, lets discuss. Perhaps some registries could be implemented without the concept of repository namespacing and always respond to requests with a global namespace regardless of the specified repository. But if they exist, documenting them here would cause a lot of confusion for end users.

Is "subject referring artifacts" a new term that needs to be defined?

I'd also avoid the v1.1 mention. The API was added in v1.1, but we have a fallback process documented in the spec that will work for older registries and could be used by tools implementing this.

Reviewer 2: (I added the suggestion from @sudo-bmitch)

Reviewer 8: was not intended to indicate registry `a` knew anything about artifacts in registry `b`  So yeah no, didn't mean to give a false impression.. Would've said all or any instead of using the `a` singular.  

Don't think a new term is needed.. an artifact that has a subject reference and is thus discoverable by referrers seems to be a subject referring artifact, separate from artifacts that do not have a subject reference field.

Of course for others reading, the fall back process, also defined in 1.1, does not use the referrers api, rather a pushed index manifest, and this is the referrers section of this doc, suppose we could change it to referrers and/or the backup processes using an index with a list of artifacts or other means. But yes we could mention other techniques other than using referrers to retrieve artifacts.

Reviewer 15: > suppose we could change it to referrers and/or the backup processes using an index with a list of artifacts or other means

I should clarify my 1.1 comment. I'd be fine with a call out to the "OCI 1.1 distribution-spec", perhaps as a link to the referrers implementation. I just didn't want to imply a requirement on a "1.1 compatible registry" for a client implementation to be able to work.

The added `mount_label` allow the kubelet to support SELinux contexts.

The kubelet will use the `mountpoint` on container creation
(by calling the `CreateContainer` RPC) to indicate the additional required volume mount ([`ContainerConfig.Mount`](https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L1102))
from the runtime. The runtime needs to ensure that mount and also manages its
lifecycle, for example to remove the bind mount on container removal.

The kubelet tracks the information about which OCI object is used by which
sandbox and therefore manages the lifecycle of them.

The proposal also considers smaller CRI changes, for example to add a list of
mounted volume paths to the `ImageStatusResponse.Image` message returned by the
`ImageStatus` RPC. This allows providing the right amount of information between
the kubelet and the runtime to ensure that no context gets lost in restart
scenarios.

The overall flow for container creation will look like this:

```mermaid
sequenceDiagram
    participant K as kubelet
    participant C as Container Runtime
    Note left of K: During pod sync
    Note over K,C: CRI

>     K->>+C: RPC: PullImage

Reviewer 9: Where does this happen? In volumemanager?

-->

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: OCIVolume
  - Components depending on the feature gate:
    - kube-apiserver (API validation)
    - kubelet (volume mount)

###### Does enabling the feature change any default behavior?

<!--
Any change of default behavior may be surprising to users or break existing
automations, so be extremely careful here.
-->

Yes, it makes the new `VolumeSource` API functional.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

<!--
Describe the consequences on existing workloads (e.g., if this is a runtime
feature, can it break the existing applications?).

Feature gates are typically disabled by setting the flag to `false` and
restarting the component. No other changes should be necessary to disable the
feature.

NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
-->

Yes, by disabling the feature gate. Existing workloads will not be affected by
the change.

To clear old volumes, all workloads using the `VolumeSource` needs to be
recreated after restarting the kubelets. The kube-apiserver does only the API

> validation whereas the kubelets serve the implementation. This means means that

Reviewer 16: nit:
s/means means/means/

## Motivation

Currently, packaging files within images often requires the inclusion of a shell, even when the sole purpose of the image is to share files among containers in a pod.
This not only increases the image size unnecessarily but also introduces potential security vulnerabilities. By supporting OCI images and artifacts
directly as a `VolumeSource`, we can streamline the process and enhance security.

### Goals

- Introduce a new `VolumeSource` type that allows mounting OCI images and/or artifacts.
- Simplify the process of sharing files among containers in a pod.
- Reduce the need for including a shell in images, thereby minimizing security risks.
- Reuse existing logic from ConfigMaps for managing file locations, garbage collection, and updates.

### Non-Goals

- This proposal does not aim to replace existing `VolumeSource` types.
- This proposal does not address other use cases for OCI images beyond file sharing among containers in a pod.

## Proposal

We propose to add a new `VolumeSource` that supports OCI images and/or artifacts. This `VolumeSource` will allow users to mount an OCI image or artifact directly into a pod,

> making the files within the image accessible to the containers without the need for a shell or additional image manipulation.

Reviewer 6: Which kinds of artifacts? Not all OCI artifacts represent filesystems.

Reviewer 6: How will the volume driver know what kind of OCI artifact it's trying to pull and mount? I know “container image” is _one_ of the options.

Reviewer 8: should we handle artifacts/images for pods shared across containers in a completely different way than artifacts/images of individual containers... hmm

Reviewer 6: If this does end up in-tree, I'd hope to see a dedicated [artifact type](https://oras.land/adopters) that means “here are files, don't try to run them, this is just data).

An example: you run an antimalware service with scanning patterns for email messages. The scanner uses a blue/green approach where the same application container image gets deployed alongside an updated scanning patterns image, _n_ times an hour. Both the application image artifact and the scanning pattern artifact are signed. The files in that signature image consist of compiled patterns that would work as a volume source, but there's no executable code in there.

(I do like the idea of making signature verification an eventual goal; that's a useful feature to leave room for).

Reviewer 17: `artifactType` or `mediaType` in the OCI manifest could give an idea of what is contained in the image and how to mount it. Filesystem type OCI images need some kind of union FS style mounting where all layers in an image are unzipped and combined to generate the final manifest whereas other artifact can be directly mounted as files

Reviewer 8: couple notes:

artifact types are defined in an IANA process.. 

runwasi(WASM) solutions are already using artifacts as a means to provide runnable code..

r we saying there may be a request for a subset of k8s artifact types? 

Will need to discuss the difference between referrers retrieved artifacts, index manifest listed artifacts, ad-hoc rootfs prepared image layers, ... how are we mapping oci to a volume and is that the right thing to do.

Reviewer 18: The image management mode of containerd is worth referring to but it's too complex if only for OCI artifacts. 
There are two core components we need, one is an OCI artifacts management like containerd image service, and another is a component to prepare a merged view of OCI volume like snapshotter

Reviewer 18: Most people use 's3fs' or 'NAS' as the backend storage for pods' data volume, which does not require downloading data locally.
﻿
For scenarios like AI/ML with a huge OCI image, full downloads can take up a lot of time and local disk space, while also adding some complexity to image management.
﻿
If the artifacts support lazy loading, then only some metadata needs to be saved locally, and there is no need to worry too much about garbage collection or layer data reuse. We only need a local cache system to save temporary data, and the management of images/artifacts can be easier.

Reviewer 6: (to me) the lazy loading story is a compelling case for allowing at least pluggable implementations; there's more than one way to lazy load an image. That could be via CSI, or we could extend CRI somehow.

Out of those two, I'd pick CSI.

Reviewer 18: I agree.
CSI is more applicable in this scenario. 
Customized CSI plugins are responsible for analyzing manifests of OCI artifacts and preparing mount points.
Then provide the data volume to the container via CSI

Reviewer 19: > Out of those two, I'd pick CSI.

Given that lazy loading image implementations already exist in the container runtimes, what leads you to prefer CSI over CRI?  (Or would you propose that a container runtime also implement CSI to provide this?)

Reviewer 6: The CSI driver could use the same code, or something close to it. However, we use CSI for all other volume types. Using CRI for a storage driver makes CRI more complex.

Reviewer 8: > > Out of those two, I'd pick CSI.
> 
> Given that lazy loading image implementations already exist in the container runtimes, what leads you to prefer CSI over CRI? (Or would you propose that a container runtime also implement CSI to provide this?)

Don't see a problem with the container runtimes providing an OCI object volume "source", that can be utilized by k8s to implement image/artifact analysis and to add oci object sourced volumes to the pod. Could do that with an extension to image service, or I suppose with a new CSI plugin type.  Would seem odd to load/manage OCI objects one way for k8s, k3s, cc, etc.. OCI pod volumes and a completely different way for container images and container referring artifacts. 

> The CSI driver could use the same code, or something close to it. However, we use CSI for all other volume types. Using CRI for a storage driver makes CRI more complex.

Not sure the ask was for an OCI storage driver.. Are you proposing using OCI image/artifact registries and proxies as remote/local storage? Or am I reading to much into the pattern proposed :-)

Reviewer 6: To me what needs clarifying is:
1. are we planning to support OCI container images as a readonly volume type, using CRI to handle those images but not to run any code
1. are we planning to support arbitrary OCI artifacts as volume sources, include OCI container images as one of potentially many kinds of artifact you can mount

If **①** then CRI feels like the better fit; that way we can use the Pod's RuntimeClass to fetch the image, in case it matters. If **②** then we can't assume CRI knows about that kind of artifact.

Reviewer 6: Either way we should document the alternative, even if we rule it out. It's good for KEPs to mention a sketch of the design(s) we didn't pick as well as the one we did.

Reviewer 8: On the arbitrary OCI artifacts note.. needs investigation/discussion.. We are talking about setting up a call to discuss next week. 

There are a number of ways artifacts have been implemented pre. and post oci 1.1 specs.

@sftm to your 1. proposal question: What does it mean to provide a mount volume into the pod's containers but not to run any code? I get that we are not expanding a pod to be an arbitrary container image/artifact with it's own command to run (yet).  Sure.  But what does it mean to mount a container image or artifact, but not to run any code there or from there? Does the volume include platforms/versions or just the one image the oci tooling would've chosen if it was being run on the node as a container by a runtime engine, does the volume include the entire index manifest tree/chain, manifest configs layers and all? Or just the layers that would otherwise makeup the container's rootfs?  

To your 2. proposal we also can't assume any CSI driver would know about all particular kinds of artifacts, registries and how to pull, parse & mount them.

minimize security risks and image size. I want to package this file in an OCI object to take advantage of OCI distribution.

#### Story 2

As a DevOps engineer, I want to package and distribute binary artifacts using OCI images and mount them directly into my Kubernetes pods,
so that I can streamline my CI/CD pipeline. I want to package this file in an OCI object to take advantage of OCI distribution.

#### Story 3

As a data scientist, MLOps engineer, or AI developer, I want to mount large language models or machine learning models in a pod alongside a model-server, so that I can efficiently serve the models
without including them in the container image. I want to package these models in an OCI object to take advantage of OCI distribution and ensure efficient model deployment.

### Notes/Constraints/Caveats (Optional)

- This enhancement assumes that the cluster has access to the OCI registry.
- The implementation must handle image pull secrets and other registry authentication mechanisms.
- Performance considerations must be taken into account, especially for large images or artifacts.

### Risks and Mitigations

> - **Security Risks:** Allowing direct mounting of OCI images introduces potential attack vectors. Mitigation includes thorough security reviews and limiting access to trusted registries.

Reviewer 20: Also need to consider how this works with image pull secrets, especially with cached images that were pulled with secrets.

    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

## Alternatives

<!--
What other approaches did you consider, and why did you rule them out? These do
not need to be as detailed as the proposal, but should include enough
information to express the idea and why it was not acceptable.

> -->

Reviewer 6: If CSI is an option for this, explain how that alternative could look.

Author: will add this

Reviewer 17: A CSI driver would be a better approach IMO

Reviewer 2: Runtimes already pull and handle images using the CRI. I'm wondering how a CSI parallel pull, auth and storage implementation is better than extending an existing solution.

Reviewer 14: I have a project that took this approach. https://github.com/converged-computing/oras-csi

Author: @vsoch very cool! is it possible to gate pulls based on signature verification & policy?

Reviewer 14: Sorry missed this! Not at the moment, but I don't see why not.

- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

The proposed enhancement adds a new `VolumeSource` to Kubernetes that supports OCI images and/or OCI artifacts.
This allows users to package files and share them among containers in a pod without including a shell in the image,
thereby reducing vulnerabilities and simplifying image creation.

While OCI images are well-supported by Kubernetes and CRI,
extending support to OCI artifacts involves recognizing additional media types, implementing custom lifecycle management,
and ensuring appropriate validation and security measures.

## Motivation

> Currently, mounting files from OCI images requires the inclusion of a shell, even when the sole purpose of the image is to deliver files

Reviewer 5: I would change the motivation to focus on OCI standards and artifacts as a way to store/distribute contents beyond runnable images because one can still create a from scratch image and add files without including a shell.

Reviewer 14: Just to be clear - ORAS means "OCI Registry as Storage" and it's main goal does exactly that.

Reviewer 2: Rephrased this section.

to be shared among containers in a pod. This not only increases the image size unnecessarily but also introduces potential security vulnerabilities.
By supporting OCI images and artifacts directly as a `VolumeSource`, we can streamline the process and enhance security.

### Goals

- Introduce a new `VolumeSource` type that allows mounting OCI images and/or artifacts.
- Simplify the process of sharing files among containers in a pod.

> - Reduce the need for including a shell in images, thereby minimizing security risks.

Reviewer 5: would drop this because it is possible to create `FROM scratch` images. This is more about consuming non runnable contents through Artifacts/images.

Reviewer 2: Yep, done.

> - Reuse existing logic from ConfigMaps for managing file locations, garbage collection, and updates.

Reviewer 5: This is an implementation detail and could be dropped I think.

Reviewer 2: Dropped.

### Non-Goals

- This proposal does not aim to replace existing `VolumeSource` types.
- This proposal does not address other use cases for OCI objects beyond file sharing among containers in a pod.

## Proposal

> We propose to add a new `VolumeSource` that supports OCI images and/or artifacts. This `VolumeSource` will allow users to mount an OCI object

Reviewer 21: Can you use Volume Populators?
One use case for Volume Populator is VM image from KubeVirt which is similar to your use case:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1495-volume-populators#vm-images

Reviewer 8: also leans into the RHCOS pattern

Reviewer 7: Hmm, the Volume Populators are definitely related, but seem a bit more block-device oriented to me?  Whereas this is much more filesystem oriented.

Reviewer 2: I agree that's a more filesystem oriented approach, not exactly sure how volume populators would fit.

Reviewer 3: Secrets and ConfigMaps are technically hardcoded volume populators for empty dir (implementation wise).  An ephemeral CSI driver can have a volume populator, right?

directly into a pod, making the files within the image accessible to the containers without the need for a shell or additional image manipulation.

### User Stories (Optional)

#### Story 1

> As a Kubernetes user, I want to share a configuration file among multiple containers in a pod without including a shell in my image, so that I can

Reviewer 5: would drop reference to shell (with same argument)

Reviewer 2: Rephrased it.

minimize security risks and image size. I want to package this file in an OCI object to take advantage of OCI distribution.

#### Story 2

As a DevOps engineer, I want to package and distribute binary artifacts using OCI images and mount them directly into my Kubernetes pods,
so that I can streamline my CI/CD pipeline. I want to package this file in an OCI object to take advantage of OCI distribution.

#### Story 3

As a data scientist, MLOps engineer, or AI developer, I want to mount large language models or machine learning models in a pod alongside a model-server, so that I can efficiently serve the models

> without including them in the container image. I want to package these models in an OCI object to take advantage of OCI distribution and ensure

Reviewer 5: separate data from executables that read/process the data.

Reviewer 2: Added

efficient model deployment.

### Notes/Constraints/Caveats (Optional)

- This enhancement assumes that the cluster has access to the OCI registry.
- The implementation must handle image pull secrets and other registry authentication mechanisms.
- Performance considerations must be taken into account, especially for large images or artifacts.

### Vocabulary: OCI Images, Artifacts, and Objects

> **OCI Image:**

Reviewer 5: Link to relevant specs.

   - A container image that conforms to the Open Container Initiative (OCI) Image Specification.
     It includes a filesystem bundle and metadata required to run a container.
   - Consists of multiple layers (each layer being a tarball), a manifest (which lists the layers), and a config file
     (which provides configuration data such as environment variables, entry points, etc.).
   - **Use Case:** Used primarily for packaging and distributing containerized applications.

> **OCI Artifact:**

Reviewer 5: Link to relevant specs.

   - Artifacts use the same image manifest and layer structure but may contain different types of data
     within those layers. The artifact manifest can have media types that differ from those in standard container images.
   - **Use Case:** Allows the distribution of non-container content using the same infrastructure and tools developed for OCI images.

**OCI Object:**
   - Umbrella term encompassing both OCI images and OCI artifacts. It represents any object that conforms to the OCI specifications for storage and distribution.

### Risks and Mitigations

- **Security Risks:** Allowing direct mounting of OCI images introduces potential attack vectors. Mitigation includes thorough security reviews and
  limiting access to trusted registries. Limiting to OCI artifacts (non-runnable content) or read-only mode may lessen the security risk.
- **Compatibility Risks:** Ensure compatibility with existing features.
- **Performance Risks:** Large images or artifacts could impact performance. Mitigation includes optimizations in the implementation and providing
  guidance on best practices for users.

## Design Details

> The new `VolumeSource` will be defined in the Kubernetes API, and the implementation will involve updating components (CRI, Kubelet, Scheduler)

Reviewer 19: Is the intent that the container runtime would be responsible for pulling the image/artifact and materializing it (e.g., handling the layer union)?  It seems like that model would allow reuse of existing logic around pull, credentials, layer backends, etc and would allow reuse of any signing integration.

Reviewer 5: Yes, that's the intent based on conversations so far. That detail should be captured here.

Reviewer 19: More thoughts on things that are involved in the existing image management area that are relevant here:
* pull
* credential management (image pull secrets and credential providers)
* garbage collection
* pod annotations/runtime handler passthrough
* lazy-loading implementations

Reviewer 5: Yes, we definitely want to retain all those features that we already use for images :+1:

Reviewer 2: I added more detail in the section. PTAL again and open new discussion points if required.

to support this source type. Key design aspects include:

- API changes to introduce the new `VolumeSource` type.

> - Modifications to the Kubelet to handle mounting OCI images and artifacts.

Reviewer 5: The runtime would be responsible for pulling and setting up the mounts and the kubelet would ask the runtime to do it. The model should be parallel to how images are handled between the kubelet and the runtime today.

Reviewer 2: I decided to only add 2 RPCs, `MountOCIObject` and `UnmountOCIObject`. Those also pull and remove without having a neeed/support for a pull policy. Would that work?

Reviewer 8: I like the parallel plan just another image.. (only these are images/artifacts on the pod and shared) I can map that out in my head.

Not sure how a no policy needed solution would work. Which policies are we discarding? never, always, if not present.

What do you mean by mount unmount apis here for the oci object(s) .. by also pull do you mean if not present, do a recheck always?    by also RM do you mean from the cache (count) so there is no need to gc by kubelet? What about disk pressure the use cases will probably cover very large artifact layers..

I could see a path where the image policy for these are hard configured on the node.

Reviewer 2: The main question for me is: Do we want to have a simple CRI interface (only mount and unmount) where the runtime should take care of the details (checking if image exists, pulling, removing when not in use any more) or should we aim for an implementation in the kubelet (like we have for image management right now)?

I'd say we move the logistics to the runtime.

Reviewer 8: I agree with the premise "runtime should take care of the details (checking if image exists, pulling, removing when not in use any more)" and need a call to discuss meaning and details.  But yes if MountOCIObject() included all of the current kubelet context for handling pod ensure image process + cache policy + pressure policy then the response to kubelet would be images/artifacts ensured see mount path and here is the status list for same.. I could also see that "long" running process as including intermediate status reports as the contents of the mounts are prepared. A one off for the pod mount(s) then transition for container images at a later date... sounds like a safe path in case people are expecting the old api to still be in use for proxy purposes.

Reviewer 22: I think precident says kubelet should be responsible for the lifecycling of the artifacts. kubelet knows disk quotas and is responsible for the state of the pods. If we treat the objects like images, and include them in ListImages(), kubelet could track them in its garbage collector.

Theoretically: if we allow OCI images, then a user could pull an image and use it for a container, as well as use it as  volume. If we didn't teach kubelet to be aware of the oci objects pulled for a volume, then it could incorrectly qualify such an image for garbage collection

Reviewer 5: Yes, it would be best let kubelet be responsible for lifecycling decisions and let the runtime service the requests. If we decide to change this split, then we should also do it for the current impl of the images as well and that should be out of scope for this KEP.

Reviewer 2: If the kubelet should take care of the lifecycle, then we I assume a breaking architectural change. Right now, the volume manager is a reconciler running in a separate go routine on kubelet startup: 

- https://github.com/kubernetes/kubernetes/blob/5bf1e955/pkg/kubelet/kubelet.go#L1634
- https://github.com/kubernetes/kubernetes/blob/5bf1e955/pkg/kubelet/volumemanager/reconciler/reconciler.go#L30-L65

The kubelet waits for the volumes to be attached within the sync loop (where also the image pull happens):

- https://github.com/kubernetes/kubernetes/blob/5bf1e955/pkg/kubelet/kubelet.go#L1945-L1952
- https://github.com/kubernetes/kubernetes/blob/5bf1e955/pkg/kubelet/kubelet.go#L1976-L1990

The whole volume manager is not aware of the CRI or the kubelet context. Means we would have to implement everything directly within the kubelet and not its volume manager and keep / extend the CRI to support OCI artifacts within the existing RPCs. 

Is that the right path forward?

Reviewer 22: I think we could teach the volume manager about the image service for this specified piece. it doesn't have to be aware of the whole sync loop as it only will use PullImage/ListImages/RemoveImage AFAIU

Reviewer 2: @haircommander yes that's true. I can elaborate on that detail as part of a PoC in the next weeks. 

@mrunalp @haircommander @mikebrow are we happy with the proposed CRI change?

Reviewer 2: Is this discussion resolved?

Reviewer 9: From storage pov, this is kinda important to decide on this. If we are going to teach volumemanager to pullimage/listimage/removeimage - then how does that integration even work? I thought it is CRI which will be responsible for this.  We definitely don't want to call CRI from volumemanager, do we?

- Enhancements to the scheduler to recognize and properly allocate resources for this new volume type.
- Handling image pull secrets and registry authentication.
- Reuse existing logic from ConfigMaps for:
  - Determining the file location on the host filesystem.
  - Managing garbage collection of unused images and artifacts.

>   - Handling updates to the OCI image or artifact, similar to how ConfigMaps update mounted files.

Reviewer 23: are these changes on the kubelet? seems related to @samuelkarp question on https://github.com/kubernetes/enhancements/pull/4642/files#r1612384741

Reviewer 2: Yes I added more detail about CRI and kubelet changes.

The following code snippet illustrates the proposed API change:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: example-pod
spec:
  volumes:
  - name: oci-volume
    oci:
      image: "example.com/my-image:latest"
  containers:
  - name: my-container
    image: busybox
    volumeMounts:
    - mountPath: /data
      name: oci-volume
```

### Kubelet and CRI support for OCI artifacts

Kubelet and the Container Runtime Interface (CRI) currently handle OCI images. To support OCI artifacts,
potential enhancements may be required:

**Extended Media Type Handling:**
   - Update Kubelet and CRI to recognize and handle new media types associated with OCI artifacts.

>    - Ensure that pulling and storing these artifacts is as efficient and secure as with OCI images.

Reviewer 23: IIUIC right now kubelet indicates the runtime to pull the images via CRI , a bit more of detail on the specific changes and defining the flow of the operations will be helpful to understand better the proposal

Reviewer 2: Added more detail, PTAL again and open new discussions if required.

Reviewer 23: I think my question is being discussed here https://github.com/kubernetes/enhancements/pull/4642/files#r1621453815 , thanks Sascha

**Lifecycling and Garbage Collection:**
   - Implement custom logic for managing the lifecycle of various OCI artifacts.

**Artifact-Specific Configuration:**
   - Introduce new configuration options or CRD to handle the unique requirements of different types of OCI artifacts.

**Validation:**
   - Extend validation and security checks to cover new artifact types.

**Storage Optimization:**

>    - Develop optimized storage solutions tailored for different artifact types,

Reviewer 5: Not sure what we mean here. Can you provide more details? I would image the runtime would be responsible for this.

Reviewer 2: @sallyom :point_up:

Author: Is this where I get to share my vision of decentralized ([IPFS](https://ipfs.tech/)) storage where users may opt-in to donate unused storage & bandwidth to address the unsustainable resource requirements of AI? 
LLMs can be chunked and distributed across the IPFS network (I've never done this, have only dreamed it so far). Has anyone proposed an IPFS CSI Plugin? 
An in-tree OCI VolumeSource with an IPFS VolumeMount provided by a CSI plugin could allow OCI volumes to be populated dynamically by a distributed IPFS system.

Reviewer 8: interesting... suppose the container runtimes could also wire in decentralized fs for cached oci object storage.. that seems like per platform/cloud level optimization details.

Reviewer 23: IPFS and decentralized filesystems have very bad performance, also the implication at the network level will be important, unless you have separate networks for the storage and the network applications ... and then you have a more complicated setup + the security concerns of the decentralized storage fs

Reviewer 5: Yes, the details of how/where the artifacts are pulled should be transparent. Kubelet only specifies the pull spec and then it is up to the runtime impl to configure how it is made available.

Reviewer 19: > up to the runtime impl to configure how it is made available

Which could then include the runtime's choice of lazy-loading implementation, pull mechanism, on-disk representation, etc.

- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

## Alternatives

### [KEP 1495: Volume Populators](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1495-volume-populators)

The volume-populators API extension allows you to populate a volume with data from an external data source when the volume is created.
This is a good solution for restoring a volume from a snapshot or initializing a volume with data from a database backup. However, it does not
address the desire to use OCI distribution, versioning, and signing for mounted data.

The proposed in-tree OCI VolumeSource provides a direct and integrated approach to mount OCI artifacts, leveraging the existing OCI
infrastructure for packaging, distribution, and security.

### Custom CSI Plugin

See [https://github.com/warm-metal/container-image-csi-driver](https://github.com/warm-metal/container-image-csi-driver)

An out-of-tree CSI plugin can provide flexibility and modularity, but there are trade-offs to consider:

>  - Complexity of managing an external CSI plugin. This includes handling the installation, configuration, and updates of the CSI driver, which adds

Reviewer 23: well, it brings the complexity to the kubelet and put it to the node maintainers :)

Reviewer 2: Yeah, …yeah.

Reviewer 24: Why was this point marked as resolved? It's a valid one and it's actually one of the reasons why CSI was introduced. If CSI is not good enough, then let's improve CSI.

Reviewer 2: > Why was this point marked as resolved?

Because this kind of discussion is not leading us into a solution and we already have stated in the alternatives section of the KEP that a custom CSI plugin is another way to solve the problem. 

What is the value of keeping this thread open?

Reviewer 6: I agree we don't need this thread for this PR. Merging this PR will produce a provisional KEP and we can revisit the CSI-vs-not discussion when we come to promote the KEP to implementable.

It _is_ important to note the alternative so that it prompts that discussion at the right time.

   an additional operational burden. For a generic, vendor-agnostic, and widely-adopted solution this would not make sense.
 - External CSI plugins implement their own lifecycle management and garbage collection mechanisms,
   yet these already exist in-tree for OCI images.

>  - Performance: There is additional overhead with an out-of-tree CSI plugin, especially in scenarios requiring frequent image pulls

Reviewer 23: can you expand more on what is this additional overhead?

   or large volumes of data.

### Advantages of In-Tree OCI VolumeSource

1. **Leverage Existing Mechanisms:**
   - **No New Data Types or Objects:** OCI images are already a core part of the Kubernetes ecosystem. Extending support for OCI artifacts, many of
     the same mechanisms will be reused. This ensures consistency and reduces complexity, as both adhere to the same OCI image format.
   - **Existing Lifecycle Management and Garbage Collection:** Kubernetes has efficient lifecycle management and garbage collection mechanisms for
	 volumes and container images. The in-tree OCI VolumeSource will utilize these existing mechanisms.

2. **Integration with Kubernetes:**

>    - **Optimal Performance:** Deep integration with the scheduler and kubelet ensures optimal performance and

Reviewer 25: https://kubernetes.io/docs/concepts/storage/volumes/#gitrepo
^ could be similar to this, deprecated though in favor on initContainer/emptyDir

We likely need a "persistent" volume source.

>    - **Optimal Performance:** Deep integration with the scheduler and kubelet ensures optimal performance and

Reviewer 18: > https://kubernetes.io/docs/concepts/storage/volumes/#gitrepo ^ could be similar to this, deprecated though in favor on initContainer/emptyDir
> 
> We likely need a "persistent" (cache?) volume source.

On-demand loading with a separate standalone cache can handle large artifact sizes very well. We have done some POC about it.

>> If the artifacts support lazy loading, then only some metadata needs to be saved locally, and there is no need to worry too much about garbage collection or layer data reuse. We only need a local cache system to save temporary data, and the management of images/artifacts can be easier.

>    - **Optimal Performance:** Deep integration with the scheduler and kubelet ensures optimal performance and

Reviewer 5: We haven't described any scheduler changes in the design details section above. One possible extension is preferring nodes that have an artifact pulled (similar to for images).

Reviewer 2: I'm not sure if we actually need scheduler changes for now. Can we drop them?

Reviewer 5: Yeah, doesn't have to part of alpha.

Reviewer 2: Removed them.

     resource management. This integration allows the OCI VolumeSource to benefit from all existing optimizations and features.
   - **Unified Interface:** Users interact with a consistent and unified interface for managing volumes, reducing the learning curve and
     potential for configuration errors.

3. **Simplified Maintenance and Updates:**
   - **Core Project Maintenance:** In-tree features are maintained and updated as part of the core project. It makes sense
     for widely-used and vendor agnostic features to utilize the core testing infrastructure, release cycles, and security updates.

### Conclusion

> The in-tree implementation of an OCI VolumeSource offers significant advantages by leveraging existing core mechanisms,

Reviewer 22: there are two advantages to in-tree: kubelet would be the one entity communicating with the runtime about objects that should exist. otherwise, the runtime would have to keep track of who requested it, and inform the kubelet not to gc an image that was pulled as a volume.

- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

The proposed enhancement adds a new `VolumeSource` to Kubernetes that supports OCI images and/or OCI artifacts.
This allows users to package files and share them among containers in a pod without including them in the main image,
thereby reducing vulnerabilities and simplifying image creation.

While OCI images are well-supported by Kubernetes and CRI,

> extending support to OCI artifacts involves recognizing additional media types, implementing custom lifecycle management,

Reviewer 2: Added as suggested.

minimize security risks and image size. 

Beside that, I want:
- to package this file in an OCI object to take advantage of OCI distribution.
- the image to be downloaded with the same credentials that kubelet using for other images.
- to be able to use image pull secrets when downloading the image if an image is from the registry that requires image pull secrets.

#### Story 2

As a DevOps engineer, I want to package and distribute binary artifacts using OCI images and mount them directly into my Kubernetes pods,
so that I can streamline my CI/CD pipeline. I want to package this file in an OCI object to take advantage of OCI distribution.

#### Story 3

As a data scientist, MLOps engineer, or AI developer, I want to mount large language models or machine learning models in a pod alongside a model-server, so that I can efficiently serve the models
without including them in the container image. I want to package these models in an OCI object to take advantage of OCI distribution and ensure
efficient model deployment. This allows to separate the data from the executables that process them.

> 

Reviewer 6: @sallyom I suggested a couple more stories - see https://github.com/kubernetes/enhancements/pull/4642#discussion_r1626349478

One is quite like the ML model serving, but I actually thought of incremental training rather than serving.

Reviewer 2: I added the story 4 to cover the malware scanner scenario.

Reviewer 6: No need to edit, this was more like an FYI to offer the stories if they are useful.

### Risks and Mitigations

- **Security Risks:** Allowing direct mounting of OCI images introduces potential attack vectors. Mitigation includes thorough security reviews and
  limiting access to trusted registries. Limiting to OCI artifacts (non-runnable content) or read-only mode may lessen the security risk.
- **Compatibility Risks:** Existing webhooks watching for the images used by the pod using some policies will need to be updated to expect the image to be specified as a `VolumeSource`.
- **Performance Risks:** Large images or artifacts could impact performance. Mitigation includes optimizations in the implementation and providing
  guidance on best practices for users.

## Design Details

The new `VolumeSource` will be defined in the Kubernetes API, and the implementation will involve updating components (CRI, Kubelet)
to support this source type. Key design aspects include:

- API changes to introduce the new `VolumeSource` type.
- Modifications to the Kubelet to handle mounting OCI images and artifacts.
- Handling image pull secrets and registry authentication.
- Reuse existing logic from ConfigMaps for:
  - Determining the file location on the host filesystem.

>   - Handling updates to the OCI image or artifact, similar to how ConfigMaps update mounted files.

Reviewer 7: This one is a really interesting topic.  IMO the fact that kube live-updates configmaps by default kind of goes against the whole immutable-infrastructure model.

I am curious what others think but I would argue that live update support should be an opt-in, [not an opt-out](https://kubernetes.io/docs/concepts/configuration/configmap/#configmap-immutable).

Reviewer 7: > ConfigMaps you can watch, whereas there's no equivalent for OCI. I can imagine it but I don't think it exists.

In OpenShift there is the [concept of an image stream](https://docs.openshift.com/container-platform/4.15/rest_api/image_apis/imagestream-image-openshift-io-v1.html), but yeah not in upstream Kube AFAIK.

Reviewer 22: BTW I don't know if we should push for this feature in alpha it seems complex

Reviewer 5: Agree that this should be left out and can be tackled later if we find use compelling uses cases or patterns that can't be solved without this.

Reviewer 15: What is the expectation of the reverse, rather than external changes pushed to the container, how do changes from the container get handled. Is the volume read-only, are writes ephemeral and discarded when the pod is removed, or are changes permitted and propagated back up somewhere?

Reviewer 7: > Is the volume read-only, 

Please yes.  Let's not repeat the sins of docker here in defaulting to a writable, ephemeral mount.  That's just an attractive trap.

The following code snippet illustrates the proposed API change:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: example-pod
spec:
  volumes:
  - name: oci-volume
    oci:
      image: "example.com/my-image:latest"
      pullPolicy: IfNotPresent
  containers:
  - name: my-container
    image: busybox
    volumeMounts:
    - mountPath: /data
      name: oci-volume
```

### Kubelet and Container Runtime Interface (CRI) support for OCI artifacts

Kubelet and the Container Runtime Interface (CRI) currently handle OCI images. To support OCI artifacts,
potential enhancements may be required:

**Extended Media Type Handling in the container runtime:**
   - Update container runtimes to recognize and handle new media types associated with OCI artifacts.

>    - Ensure that pulling and storing these artifacts is as efficient and secure as with OCI images.

Reviewer 26: if the volume source is updated to use a new released model artifact, then these images which were mounted with old artifact should unmount the old one first, and re-mount the new artifact files. 

how to deal with the case if the images are accessing these files duing mount/un-mount?

Reviewer 2: Is this discussion referring to https://github.com/kubernetes/enhancements/pull/4642#discussion_r1626363466?

Reviewer 6: With a CSI approach, we could offer in-place updates (where `inotify` and friends would fire). If the container runtime does the image pull, we don't really have that option.

Reviewer 2: > in-place updates
We could during sync pod, but I assume we left that out intentional as discussed earlier.

**Lifecycling and Garbage Collection:**
   - Reuse the existing kubelet logic for managing the lifecycle of OCI objects.

**Artifact-Specific Configuration:**
   - Introduce new configuration options to handle the unique requirements of different types of OCI artifacts.

> 

Reviewer 8: ```suggestion
**Artifacts as Subject Referrers:**
   - Introduce new refer to image and filter for artifact type criterion/options for to be mounted artifact(s).
   - Certain types of OCI artifacts include a subject reference.  That reference identifies the artifact/image for which this artifact refers. For example a signature artifact could refer to a platform index, for certifying the platform images, or to an SBOM artifact that refers to a platform matched image. These artifacts may or may not be located on the same registry/repository. The new referrers api allows for retrieving these artifacts.
```

Reviewer 8: or we can push that to r.next as this pattern is fairly new and needs more api changes / validation and roll out.

Reviewer 2: I add it because I think it's important. Don't see it in alpha, though.

life-cycle of them. This means that some parts of the existing kubelet code can
be reused, for example:

- The logic how to ensure that an image exists on the node:
  https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/images/image_manager.go#L102
- The retrieval of available secrets for a pod:
  https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/kubelet_pods.go#L988

Specifying a `pullPolicy` be supported in the same way as for the existing
`imagePullPolicy` API.

For registry authentication purposes the same logic will be used as for the
container image.

#### CRI

The CRI API is already capable of managing container images [via the `ImageService`](https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L146-L161).
Those RPCs will be re-used for managing OCI artifacts, while the [`ImageSpec`](https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L798-L813)
will be extended to mount the OCI object to a local path:

```protobuf

// ImageSpec is an internal representation of an image.
message ImageSpec {
    // …

    // Absolute local path where the image/artifacts should be mounted to.

>     string mount_path = 20;

Reviewer 19: ```suggestion
    string mountpoint = 20;
```

This is the more standard terminology, see [`mount(8)`](https://man7.org/linux/man-pages/man8/mount.8.html).

}
```

This allows to re-use the existing kubelet logic for managing the OCI objects,
with the caveat that the new `VolumeSource` won't be isolated in a dedicated
plugin as part of the existing [volume manager](https://github.com/kubernetes/kubernetes/tree/6d0aab2/pkg/kubelet/volumemanager).

#### Container Runtimes

> Container runtimes need to support the new `mount_path` field, otherwise the

Reviewer 19: ```suggestion
Container runtimes need to support the new `mountpoint` field, otherwise the
```

> feature cannot be used. The kubelet will verify if the mount path actually

Reviewer 19: ```suggestion
feature cannot be used. The kubelet will verify if the mountpoint actually
```

exists on disk to check the feature availability, because Protobuf will strip
the field in a backwards compatible way for older runtimes. Pods using the new
`VolumeSource` combined with a not supported container runtime version will fail
to run on the node.

> For security reasons, volume mounts should be done using the [`noexec`] flag.

Reviewer 19: ```suggestion
For security reasons, volume mounts should set the [`noexec`] and `ro` options by default.
```

Opening this as a separate discussion, but it would be reasonable to make these [read-only](https://github.com/kubernetes/enhancements/pull/4642/files#r1626748582) by default.

Reviewer 2: Yep, done

The volume-populators API extension allows you to populate a volume with data from an external data source when the volume is created.
This is a good solution for restoring a volume from a snapshot or initializing a volume with data from a database backup. However, it does not
address the desire to use OCI distribution, versioning, and signing for mounted data.

The proposed in-tree OCI VolumeSource provides a direct and integrated approach to mount OCI artifacts, leveraging the existing OCI
infrastructure for packaging, distribution, and security.

### Custom CSI Plugin

See [https://github.com/warm-metal/container-image-csi-driver](https://github.com/warm-metal/container-image-csi-driver)

An out-of-tree CSI plugin can provide flexibility and modularity, but there are trade-offs to consider:

 - Complexity of managing an external CSI plugin. This includes handling the installation, configuration, and updates of the CSI driver, which adds
   an additional operational burden. For a generic, vendor-agnostic, and widely-adopted solution this would not make sense.

>  - Supporting the image pull secrets as well as credentials provider will be tricky and needs to be reimplemented with the separate API calls.

Reviewer 24: If current handling of secrets in CSI is insufficient, we should improve that instead of introducing a separate volume plugin that will need to solve the same issues, just in-tree.

 - External CSI plugins implement their own lifecycle management and garbage collection mechanisms,
   yet these already exist in-tree for OCI images.
 - Performance: There is additional overhead with an out-of-tree CSI plugin, especially in scenarios requiring frequent image pulls
   or large volumes of data.

> ### Advantages of In-Tree OCI VolumeSource

Reviewer 24: Most of the in-tree advantages listed here can be done with an external CSI driver. To me it seems this proposal tries to cut corners, because implementing it in-tree may look easier, but IMO it will not pay off in the long run.

We introduced CSI as an extension mechanism for all volumes, incl. ephemeral ones. If CSI is not good enough, then let's fix CSI and allow anyone to write similar drivers - OCI is not the only format in the world, even if it's currently popular.

1. **Leverage Existing Mechanisms:**
   - **No New Data Types or Objects:** OCI images are already a core part of the Kubernetes ecosystem. Extending support for OCI artifacts, many of

>      the same mechanisms will be reused. This ensures consistency and reduces complexity, as both adhere to the same OCI image format.

Reviewer 24: There is no reason why an external CSI driver could not use "OCI artifacts" and "adhere to the same OCI image format".

   - **Existing Lifecycle Management and Garbage Collection:** Kubernetes has efficient lifecycle management and garbage collection mechanisms for

> 	 volumes and container images. The in-tree OCI VolumeSource will utilize these existing mechanisms.

Reviewer 24: This is actually the only one argument for in-tree volume plugin that I can agree with. It's kubelet who performs image garbage collection and it must know that an image is used as a volume in a running Pod.

I'd appreciate some effort to find a solution that can work with out-of-tree CSI driver. https://kubernetes.io/docs/concepts/architecture/garbage-collection/#containers-images says that cadvisor is somehow used there - can it report images used as volumes? Can a CSI driver or anything else running on a node "hold" an image from garbage collector?

2. **Integration with Kubernetes:**
   - **Optimal Performance:** Deep integration with the scheduler and kubelet ensures optimal performance and

>      resource management. This integration allows the OCI VolumeSource to benefit from all existing optimizations and features.

Reviewer 24: How exactly need OCI image volumes interact with scheduler? What existing (scheduler?) optimizations and features will it use? Why is it not possible with CSI drivers?

Reviewer 6: The scheduler needs to know that some node-local storage would be used up on the node where the Pod is running (for the unpacked image).

Of course, **that's not entirely true**: the CSI driver might implement some other mechanism - a memory backed filesystem for example, or a userspace filesystem backed by on demand network fetches, or the unpacked filesystem could be placed onto network storage. We shouldn't assume that local unpacking the way Docker does it is the only viable option.

   - **Unified Interface:** Users interact with a consistent and unified interface for managing volumes, reducing the learning curve and

>      potential for configuration errors.

Reviewer 24: Users already know how to handle CSI volumes, they've been there since 2019.

3. **Simplified Maintenance and Updates:**
   - **Core Project Maintenance:** In-tree features are maintained and updated as part of the core project. It makes sense

>      for widely-used and vendor agnostic features to utilize the core testing infrastructure, release cycles, and security updates.

Reviewer 24: We have widely-used and vendor agnostic CSI drivers under Kubernetes umbrella - NFS, SMB, and shared-secrets. They use the core testing infrastructure. They can have different cycles and security updates. if they're not good enough, we should fix those processes for all non-in-tree components, not merge everything back to k/k.

Reviewer 6: “Core Project Maintenance” can sometimes read like: if we get this merged, someone will _have_ to maintain it.

any content using OCI registries. This allows the project to grow into use cases
which go beyond running particular images.

### Goals

- Introduce a new `VolumeSource` type that allows mounting OCI images and/or artifacts.
- Simplify the process of sharing files among containers in a pod.

### Non-Goals

- This proposal does not aim to replace existing `VolumeSource` types.
- This proposal does not address other use cases for OCI objects beyond file sharing among containers in a pod.

## Proposal

We propose to add a new `VolumeSource` that supports OCI images and/or artifacts. This `VolumeSource` will allow users to mount an OCI object
directly into a pod, making the files within the image accessible to the containers without the need to include them in the main image and to be able to host them in OCI compatible registries.

### User Stories (Optional)

#### Story 1

As a Kubernetes user, I want to share a configuration file among multiple containers in a pod without including the file in my main image, so that I can
minimize security risks and image size. I want to package this file in an OCI object to take advantage of OCI distribution.

> 

Reviewer 10: ```suggestion
I want the image to be downloaded with the same credentials that kubelet using for other images. I also want to be able to use image pull secrets when downloading the image if an image is from the registry that requires image pull secrets.
```

#### Story 2

> As a DevOps engineer, I want to package and distribute binary artifacts using OCI images and mount them directly into my Kubernetes pods,

Reviewer 10: for the binary we need to make sure that the host architecture is matching. This is part of a plan, correct?

Reviewer 6: :thought_balloon: I wonder if, _by default_, we should mount this kind of volume as `noexec`. I'd want that default.

Reviewer 2: Yes, that should be done by the runtime on pull.

Reviewer 2: Added a note about mounting with `noexec` within runtimes.

Reviewer 8: We have a couple ways to make sure the artifact is "matching" wrt platform host arch/version etc. 

There is a new referrers api in the oci 1.1 spec that would allow us the opportunity to get a list of / filter artifacts that are referring to a platform image.. (if the registry and client supports the referrers api).  Rough steps .. pull platform index, match platform image, request artifacts referring to matched platform image, match requested artifact from this new k8s api.  We can discuss other patterns that do not rely on the referrers api on a per registry basis, and not all "1.0" oci registries support artifacts in the same way yet. 

We've had some discussions about a-priori pushing oci index content that includes artifacts ordered by platform. That would change up the ^ sequence to pulling the set of referred / tagged manifests and finding/matching the artifacts within that set. 

@sudo-bmitch

**OCI Artifact ([guidance](https://github.com/opencontainers/image-spec/blob/main/artifacts-guidance.md)):**
   - An artifact describes any content that is stored and distributed using the OCI image format.
     It includes not just container images but also other types of content like Helm charts, WASM modules, machine learning models, etc.
   - Artifacts use the same image manifest and layer structure but may contain different types of data
     within those layers. The artifact manifest can have media types that differ from those in standard container images.
   - **Use Case:** Allows the distribution of non-container content using the same infrastructure and tools developed for OCI images.

**OCI Object:**
   - Umbrella term encompassing both OCI images and OCI artifacts. It represents any object that conforms to the OCI specifications for storage and distribution.

### Risks and Mitigations

- **Security Risks:** Allowing direct mounting of OCI images introduces potential attack vectors. Mitigation includes thorough security reviews and
  limiting access to trusted registries. Limiting to OCI artifacts (non-runnable content) or read-only mode may lessen the security risk.

> - **Compatibility Risks:** Ensure compatibility with existing features.

Reviewer 10: ```suggestion
- **Security Risks:** Existing webhooks watching for the images used by the pod using some policies will need to be updated to expect the image to be specified as a VolumeSource.
- **Compatibility Risks:** Ensure compatibility with existing features.
```

**Validation:**
   - Extend validation and security checks to cover new artifact types.

**Storage Optimization:**
   - Develop optimized storage solutions tailored for different artifact types,
     potentially integrating with existing storage solutions or introducing new mechanisms.

#### kubelet

Some parts of the existing kubelet code can be reused, for example:

- The logic how to ensure that an image exists on the node:
  https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/images/image_manager.go#L102
- The retrieval of available secrets for a pod:
  https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/kubelet_pods.go#L988

We consider to refactor the logic into re-usable bits for both, the kubelet as
well as the volume plugin, because volume plugins run in a dedicated go routine
at the beginning of the kubelet initialization:

https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/kubelet.go#L1633-L1634

Specifying a pull policy will not be supported in the alpha implementation of the
feature and will align to `PullAlways`.

For registry authentication purposes the same credentials will be used as for

> the running pod, if available.the

Reviewer 10: ```suggestion
the running pod, if available.
```

#### CRI

The CRI is already capable of pulling images [via the `PullImage` RPC](https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L154).
This does not necessarily include OCI artifacts and is dependent on the used
container runtime (CRI-O, containerd). Beside that, volume plugins usually
require a mount from a local directory. This means we extend the CRI protocol by
adding a two new methods as well as their request and response types:

```protobuf

> // ImageService defines the public APIs for managing images.

Reviewer 10: should we pass the runtime handler and pod annotations to align with what we do for regular images

- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

## Alternatives

### [KEP 1495: Volume Populators](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1495-volume-populators)

The volume-populators API extension allows you to populate a volume with data from an external data source when the volume is created.
This is a good solution for restoring a volume from a snapshot or initializing a volume with data from a database backup. However, it does not
address the desire to use OCI distribution, versioning, and signing for mounted data.

The proposed in-tree OCI VolumeSource provides a direct and integrated approach to mount OCI artifacts, leveraging the existing OCI
infrastructure for packaging, distribution, and security.

### Custom CSI Plugin

See [https://github.com/warm-metal/container-image-csi-driver](https://github.com/warm-metal/container-image-csi-driver)

An out-of-tree CSI plugin can provide flexibility and modularity, but there are trade-offs to consider:

>  - Complexity of managing an external CSI plugin. This includes handling the installation, configuration, and updates of the CSI driver, which adds

Reviewer 10: ```suggestion
  - Supporting the image pull secrets as well as credentials provider will be tricky and needs to be reimplemented with the separate API calls
  - Complexity of managing an external CSI plugin. This includes handling the installation, configuration, and updates of the CSI driver, which adds
```

Reviewer 2: Added this as separate point if that makes sense.

- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

The proposed enhancement adds a new `VolumeSource` to Kubernetes that supports OCI images and/or OCI artifacts.
This allows users to package files and share them among containers in a pod without including them in the main image,

> thereby reducing vulnerabilities and simplifying image creation.

Reviewer 27: Adding extra package sourcing would result in an increase in attack surface. Stating as a conclusion that `VolumeSource`s will reduce vulnerabilities might therefore not be the case. Unless`VolumeSource`s isn't used in conjunction with image repos. A combination of both might result in higher vulnerability.

Author: Thanks for bringing this up - I agree the Summary is not as clearly stated as it should be - I'll update it - my thinking is more this: 
```
Introducing a new VolumeSource does not inherently reduce vulnerabilities, but design choices around read-only,
non-executable mounts, and the decoupling of application and data can contribute to a more secure overall system.
The goal is to provide a secure, modular, and flexible way to manage files (most importantly today but not limited to 
model files), that aligns with best practices for container security. For example, a model-server can switch between
different models without altering model-server container. This flexibility simplifies updates and allows for quick
changes without requiring a full container rebuild.
```

**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

The proposed enhancement adds a new `VolumeSource` to Kubernetes that supports OCI images and/or OCI artifacts.
This allows users to package files and share them among containers in a pod without including them in the main image,
thereby reducing vulnerabilities and simplifying image creation.

While OCI images are well-supported by Kubernetes and CRI,
extending support to OCI artifacts involves recognizing additional media types, implementing custom lifecycle management,
and ensuring appropriate validation and security measures.

## Motivation

Supporting OCI images and artifacts directly as a `VolumeSource` allows
Kubernetes to focus on OCI standards as well as allows to store and distribute
any content using OCI registries. This allows the project to grow into use cases
which go beyond running particular images.

### Goals

> - Introduce a new `VolumeSource` type that allows mounting OCI images and/or artifacts.

Reviewer 15: Images with layers have a well known media type for the layers, and defined behaviors for expanding the collection of tar layers into an overlay filesystem. But artifacts can have any media type, not necessarily a tar with optional compression. How would unknown media types be handled, and would it be possible to extract a single file directly stored as a single layer with a specified media type?

For example, wasm style artifacts may include the runnable binary directly as a layer, or a ML model may have their own binary format for packaging their model that doesn't require additional wrapping with tar or gzip.

Reviewer 8: Brandon, this VolumeSource type is for the kubernetes API, vs an artifact type.  I agree, may have mentioned, but not to sufficient detail yet, the issue with IANA types possibly having their own definitions for if and how they should be "mounted."

Reviewer 7: > . How would unknown media types be handled, and would it be possible to extract a single file directly stored as a single layer with a specified media type?

Isn't it more that we would need to explicitly define media types that *are* handled?  I am not seeing much in the text about the media type handling beyond
>  Update container runtimes to recognize and handle new media types associated with OCI artifacts.

But I think this needs to be explicitly specified.

I think there's going to be a split between:

"I want to embed this single file at the target path" which would map to the AI model case, as well as other things like "I want to inspect this ISO image which is stored as an OCI artifact".
vs
"I have a filesystem tree" use case....which is actually a lot like what ConfigMaps are.

Reviewer 15: For the media types that are handled, make sure that also covers the scenario with multiple layers. There are some assumptions when you go from one tar file to multiple tar files that you'll want to document.

For the unknown media types, I think the options are to either reject, ignore/skip, or allow the user to map specific media types to specific files. The latter would also need to handle multiple layer scenarios.

This needs to avoid any design that would change behavior as media types change from unknown to known in later releases (you don't want a file to suddenly become a filesystem in a new release). So a user supplied mapping could override any known media type actions.

Reviewer 28: Would there be a benefit to supporting a media type + an artifact type (+ platform)?

Suppose you have an Image Index which references multiple Image Manifests which themselves reference multi-layer blobs. Each of these Image Manifests could potentially have unique platforms/artifactTypes representing different configurations.

Distributing artifacts in this way would be beneficial as it would enable the grouping of conceptually similar artifacts together. This would be potentially beneficial from a storage perspective as well because it would enable caching of each individual layers while still enabling an entire filesystem to be represented.

Reviewer 26: Sometimes the owner/author of AI/ML model is different from the image, and the model may be updated independent from the image. model as a shared layer will lead to challenge for online training and model updating, if layer is updated, the images(which share the model) have to update too.

Reviewer 2: I intentionally left open / removed the media type discussion because:

- there is no standard yet and creating one takes time (tried that for a seccomp profile a while ago).
- runtimes can still decide which media types to accept or reject
- we can improve this KEP and the API later on if a standard exists

- Simplify the process of sharing files among containers in a pod.

### Non-Goals

- This proposal does not aim to replace existing `VolumeSource` types.
- This proposal does not address other use cases for OCI objects beyond file sharing among containers in a pod.

## Proposal

We propose to add a new `VolumeSource` that supports OCI images and/or artifacts. This `VolumeSource` will allow users to mount an OCI object
directly into a pod, making the files within the image accessible to the containers without the need to include them in the main image and to be able to host them in OCI compatible registries.

### User Stories (Optional)

#### Story 1

As a Kubernetes user, I want to share a configuration file among multiple containers in a pod without including the file in my main image, so that I can
minimize security risks and image size. 

Beside that, I want:
- to package this file in an OCI object to take advantage of OCI distribution.
- the image to be downloaded with the same credentials that kubelet using for other images.
- to be able to use image pull secrets when downloading the image if an image is from the registry that requires image pull secrets.

#### Story 2

> As a DevOps engineer, I want to package and distribute binary artifacts using OCI images and mount them directly into my Kubernetes pods,

Reviewer 15: I worry this would get overused instead of creating an image with a single added layer.

Reviewer 8: nod.. some of the language has been artifacts/images.. 

```suggestion
As a DevOps engineer, I want to package and distribute binary artifacts using OCI image and distribution specification standards and mount them directly into my Kubernetes pods,
```

Though it's not clear if that means we need another story for docker manifest types.. maybe best to just say Docker/OCI here.

Reviewer 22: > I worry this would get overused instead of creating an image with a single added layer.

I think that is the advantage here: Sharing single files/layers across different images without having to sync the base layers. it's a workaround on oci image spec not being smarter about shared files

Reviewer 15: > Sharing single files/layers across different images without having to sync the base layers.

Is "sync the base layers" a reference to coping layers between registries and repositories? If the volume source and container image are in separate repositories, that could have authentication impacts (credentials able to pull an image but not the volume source).

Reviewer 22: not quite: two different containers sharing the same file but otherwise don't share anything else have two options, share a same base layer with that file (not super idiomatic, as I think lots of non-multi-stage-built images are built starting with package installation, which in this case would be separate), or sharing them with a volume. but a volume doesn't give the utility and tooling of oci distribution spec. I find this to be the best of both worlds

Reviewer 15: When in the same repository, a common layer can be easily added to multiple images. Much easier, and more portable, than adding a k8s specific volume.

Reviewer 22: I may be speaking out of my depth here, but if they're not explicitly sharing a base layer, will the storage for that layer be actually shared in the containerd/cri-o level? will the sha for that layer be the same so it's not pulled twice?

Reviewer 15: The sha for a layer can be the same across multiple images, and when that happens, most runtimes will dedup the layers, skipping pulls for content it already has.

Reviewer 6: > I think that is the advantage here: Sharing single files/layers across different images without having to sync the base layers. it's a workaround on oci image spec not being smarter about shared files

To me this is about separating code from data.

You might also want to use a public image but with your own data. For example, you use a public image for a malware scanner and mount in a volume of private (commercial) malware signatures. This KEP gives you a way to load those signatures without baking your own combined image (which might not be allowed by the copyright on the public image), and they'll work regardless of the OS or version of the scanning software.

It'll work with other kinds of read-only data; for example, you could have some ML retraining code (eg Python) and a base model (half a terabyte of sparse matrices). Iterating on the Python shouldn't mean re-pulling the base model.

Reviewer 6: If we're talking about combining _code_ with code, is that something we'd like to explicitly encourage?

Reviewer 7: > If we're talking about combining code with code, is that something we'd like to explicitly encourage?

There's multiple levels to this; when talking about "code" there's actually a somewhat fuzzy line in some cases between "configuration" and "code".  But an example *strong* dividing line is as soon as you have architecture (aarch64/x86_64/s390x/...) dependent stuff it's *clearly* code (and gets close to OCI containers obviously).

Whereas I think the primary use cases this starts from are much closer to architecture-independent data.

Reviewer 2: The story seems to have multiple shapes, but generally I do not see any blockers in the implementation proposal about them. I rephrase the story as @mikebrow suggested. 

Is there anything left open as part of this discussion?

Reviewer 29: Hi everyone, I'm proposing to add a new MediaType to the OCI image spec that will only hold raw data files, for optimization of large files in the image, please take a look.
https://github.com/opencontainers/image-spec/pull/1197

- **Performance Risks:** Large images or artifacts could impact performance. Mitigation includes optimizations in the implementation and providing
  guidance on best practices for users.

## Design Details

The new `VolumeSource` will be defined in the Kubernetes API, and the implementation will involve updating components (CRI, Kubelet)
to support this source type. Key design aspects include:

- API changes to introduce the new `VolumeSource` type.
- Modifications to the Kubelet to handle mounting OCI images and artifacts.
- Handling image pull secrets and registry authentication.
- Reuse existing logic from ConfigMaps for:
  - Determining the file location on the host filesystem.
  - Managing garbage collection of unused images and artifacts.
  - Handling updates to the OCI image or artifact, similar to how ConfigMaps update mounted files.

The following code snippet illustrates the proposed API change:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: example-pod
spec:
  volumes:
  - name: oci-volume
    oci:

>       image: "example.com/my-image:latest"

Reviewer 23: @tallclair I wonder if using have some implications 🤔

Reviewer 22: since we're now PullAlways for everything, I think we don't need to handle that. eventually, if we add more pull policies, we should treat the interpretation of these digests the same as an image.

  containers:
  - name: my-container
    image: busybox
    volumeMounts:
    - mountPath: /data
      name: oci-volume
```

### Kubelet and Container Runtime Interface (CRI) support for OCI artifacts

Kubelet and the Container Runtime Interface (CRI) currently handle OCI images. To support OCI artifacts,
potential enhancements may be required:

**Extended Media Type Handling:**

>    - Update Kubelet and CRI to recognize and handle new media types associated with OCI artifacts.

Reviewer 22: I don't know if the kubelet needs to be aware of media type

Reviewer 5: Agree, especially, if these volumes are always `noexec`.

Reviewer 6: The runtime might need to report back to the kubelet what artifact type it has pulled, and that the runtime [does|doesn't] know how to turn that into a mountable volume.

Reviewer 2: I changed the point to refer to the runtime implementation, which seems fair from my point of view.

   - Ensure that pulling and storing these artifacts is as efficient and secure as with OCI images.

**Lifecycling and Garbage Collection:**
   - Implement custom logic for managing the lifecycle of various OCI artifacts.

**Artifact-Specific Configuration:**

>    - Introduce new configuration options or CRD to handle the unique requirements of different types of OCI artifacts.

Reviewer 23: how CRDs will work?
this raises another question, is this option available for disconnected kubelets?

Reviewer 6: > disconnected kubelets

Do you mean “can static Pods use this volume source”?

Reviewer 22: I feel like they should be treated the same as hostPath (I'm not sure if those work with static pods off the top of my head)

Reviewer 6: Static Pods **can** use `hostPath` (it's quite common). IIRC they can't use `imagePullSecrets` and a few things like that, though.

Reviewer 2: I removed the note to the CRD, I think it's a bit out of scope for now.

**Validation:**
   - Extend validation and security checks to cover new artifact types.

**Storage Optimization:**
   - Develop optimized storage solutions tailored for different artifact types,
     potentially integrating with existing storage solutions or introducing new mechanisms.

#### kubelet

Some parts of the existing kubelet code can be reused, for example:

- The logic how to ensure that an image exists on the node:
  https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/images/image_manager.go#L102
- The retrieval of available secrets for a pod:
  https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/kubelet_pods.go#L988

We consider to refactor the logic into re-usable bits for both, the kubelet as
well as the volume plugin, because volume plugins run in a dedicated go routine
at the beginning of the kubelet initialization:

https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/kubelet.go#L1633-L1634

> Specifying a pull policy will not be supported in the alpha implementation of the

Reviewer 23: I think it should be discussed anyway, it seems an important detail, otherwise it will cause a performance impact, no?

Reviewer 8: using Pull Always breaks Never pull (disconnected - ignore `latest` tag)

If Not Present is a policy that significantly reduces time to start a pod, some registry providers count checking manifests of an image as a full pull for the image. The time cost of making an authorized connection and checking the manifest tree for local existence of all sha's is, significant, esp. when the use count hits the user's limit at the registry.

Reviewer 22: though: shouldn't the runtimes be optimizing and not repulling bits they already have? they have to do the auth check, and it does now rely on the registry always being up (which was something we steered away from in https://github.com/kubernetes/enhancements/issues/2535 ) but at least the runtimes should optimize so we're not actually pulling always if the bits are already on disk at the right location

Reviewer 6: > shouldn't the runtimes be optimizing and not repulling bits they already have

Yes, that's typical.

Reviewer 8: > though: shouldn't the runtimes be optimizing and not repulling bits they already have? they have to do the auth check, and it does now rely on the registry always being up (which was something we steered away from in #2535 ) but at least the runtimes should optimize so we're not actually pulling always if the bits are already on disk at the right location

Yes, of course. 

My point was about the time to auth connect and checking the manifests, by request, when the image is already present. There is usually no charge for "connect" .. the charge is for the manifest checking, even though the cost is expected to minimal with existing size limits on the number of annotations/size of config etc. Pulling the layer blobs is not needed when their manifest referenced SHAs are in cache.

Thus not having a policy at all for these new pulls, begs the question. Though with the appropriate container runtime config (a never pull disconnected state switch) and using the alg from the ensure secret KEP we can presume pull if not present and ensure at the container runtime.  This would be my suggestion.

Reviewer 8: Fair and if we go back to the prior approach again.. note the nit to allow the CRI integrations to provide a suitable and safe alternative config.

feature and will align to `PullAlways`.

For registry authentication purposes the same credentials will be used as for
the running pod, if available.

#### CRI

The CRI is already capable of pulling images [via the `PullImage` RPC](https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L154).
This does not necessarily include OCI artifacts and is dependent on the used
container runtime (CRI-O, containerd). Beside that, volume plugins usually
require a mount from a local directory. This means we extend the CRI protocol by

> adding a two new methods as well as their request and response types:

Reviewer 22: an alternative (especially if kubelet handles the lifecycle of the pods) is to extend the PullImage rpc to contain a target path field, and the CRI pulls to that path if it's specified. removal can implicitly include unmount if the kubelet includes the path variable as well. That could reduce duplicatio between the 4 RPCs

Reviewer 2: Yes, see the discussion at: https://github.com/kubernetes/enhancements/pull/4642#discussion_r1625550795

```protobuf
// ImageService defines the public APIs for managing images.
service ImageService {
    // MountOCIObject pulls and mounts any OCI image or artifact to a local
    // directory on disk
    rpc MountOCIObject(MountOCIObjectRequest) returns (MountOCIObjectResponse) {}

    // UnmountOCIObject unmounts and also removes any OCI image or artifact.
    rpc UnmountOCIObject(UnmountOCIObjectRequest) returns (UnmountOCIObjectResponse) {}

    /// …
}

message MountOCIObject {
    // Reference to be pulled.
    string object = 1;

    // Authentication configuration for pulling the object.
    AuthConfig auth = 2;

    // Local path to be mounted.
    string path = 3;

    // Corresponding to:
    // https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L809-L812
    string runtime_handler = 4;

    // Corresponding to:
    // https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L802-L805
    map<string, string> annotations = 5;
}

message MountOCIObjectResponse {
    // Reference to the object in use.
    string id = 1;
}

message UnmountOCIObjectRequest {
    // Identifier of the object to remove.

>     string id = 1;

Reviewer 22: should we include path as well, in case kubelet decides to pull to different locations?

Reviewer 2: Now part of the new API.

## Proposal

We propose to add a new `VolumeSource` that supports OCI images and/or artifacts. This `VolumeSource` will allow users to mount an OCI object
directly into a pod, making the files within the image accessible to the containers without the need to include them in the main image and to be able to host them in OCI compatible registries.

### User Stories (Optional)

#### Story 1

As a Kubernetes user, I want to share a configuration file among multiple containers in a pod without including the file in my main image, so that I can
minimize security risks and image size. 

Beside that, I want:
- to package this file in an OCI object to take advantage of OCI distribution.
- the image to be downloaded with the same credentials that kubelet using for other images.
- to be able to use image pull secrets when downloading the image if an image is from the registry that requires image pull secrets.

#### Story 2

As a DevOps engineer, I want to package and distribute binary artifacts using OCI image and distribution specification standards and mount them directly into my Kubernetes pods,

> so that I can streamline my CI/CD pipeline. I want to package this file in an OCI object to take advantage of OCI distribution.

Reviewer 30: Can we highlight the specific advantages of using this for CI/CD pipelines?

Reviewer 2: Yep, added them. :+1:

#### Story 3

As a data scientist, MLOps engineer, or AI developer, I want to mount large language models or machine learning models in a pod alongside a model-server, so that I can efficiently serve the models
without including them in the container image. I want to package these models in an OCI object to take advantage of OCI distribution and ensure

> efficient model deployment. This allows to separate the data from the executables that process them.

Reviewer 30: "data" might be controversial since some may think that models are code. For this specific use case, perhaps we can clarify this to be "model specifications/content"?
